### PR TITLE
mgr/test_orchestrator: Allow initializing dummy data

### PR DESCRIPTION
**Pull request by bk201**
- Add test_orchestrator load_data command
- Add QA tests for loading data

Fixes: https://tracker.ceph.com/issues/41151

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: qa/tasks/mgr/test_orchestrator_cli.py [hunk 1]
```diff
@@ -153,3 +160,78 @@ def test_progress(self):
...
+        self.assertEqual(ret, errno.EINVAL)
+
+        # test invalid input file: missing key
+        json_str = '{ "inventory": [{"devices": []}] }'
+        ret = self._test_orchestrator_cmd_result('load_data', '-i', '-', stdin=json_str)
+        self.assertEqual(ret, errno.EINVAL)
+
+        # disable test_orchestrator to reload data for other tests
+        self.mgr_cluster.mon_manager.raw_cluster_cmd('mgr', 'module', 'disable',
+                                                     'test_orchestrator')
```
[the remark this answers is not in the document]
Comment (bk201) [reply qa/tasks/mgr/test_orchestrator_cli.py]: Thanks! Fixed in https://github.com/ceph/ceph/pull/29595/commits/da33500b25bd29fbd9bccdb931c6ccc7492c817f

**Comment by tchaikov** [comment]
@bk201 could you squash these two commits into a single one?

**Comment by bk201** [comment]
> @bk201 could you squash these two commits into a single one?

Fixed. Thanks!

**Comment by tchaikov** [comment]
```
 PoolDetailsComponent › Pool details tabset › current active status of tabs should not change when selection is same with previour selection

    HttpErrorResponse: Http failure response for api/summary: 0 Unknown Error

    Error: 

      72 |     });
      73 | 
    > 74 |     it('current active status of tabs should not change when selection is same with previour selection', () => {
         |     ^
      75 |       fixture.detectChanges();
      76 |       const tabs = poolDetailsComponent.tabsetChild.tabs;
      77 |       expect(tabs[0].active).toBeTruthy();

      at new Spec (node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
      at env.(anonymous function) (node_modules/jest-preset-angular/zone-patch/index.js:84:27)
      at src/app/ceph/pool/pool-details/pool-details.component.spec.ts:74:5
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:391:26)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone.js:150:43)
      at Suite.<anonymous> (node_modules/jest-preset-angular/zone-patch/index.js:40:21)
      at env.(anonymous function) (node_modules/jest-preset-angular/zone-patch/index.js:69:27)
      at src/app/ceph/pool/pool-details/pool-details.component.spec.ts:36:3
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:391:26)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone.js:150:43)
      at Suite.<anonymous> (node_modules/jest-preset-angular/zone-patch/index.js:40:21)
      at env.(anonymous function) (node_modules/jest-preset-angular/zone-patch/index.js:69:27)
      at Object.<anonymous> (src/app/ceph/pool/pool-details/pool-details.component.spec.ts:12:1)
```

is this something we should take a look at?

File: src/pybind/mgr/orchestrator.py [hunk 2]
```diff
@@ -51,6 +51,12 @@ class OrchestratorValidationError(OrchestratorError):
     """
 
 
+class OrchestratorDeserializeError(OrchestratorError):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]: Yeah, previously, I used `orchestrator.OrchestratorValidationError` for that.

File: src/pybind/mgr/orchestrator.py [hunk 3]
```diff
@@ -906,6 +916,17 @@ def __init__(self, name, devices):
...
         return {'name': self.name, 'devices': [d.to_json() for d in self.devices]}
 
+    @classmethod
+    def from_json(cls, data):
+        try:
+            name = data['name']
+            devices = [InventoryDevice.from_json(device)
+                       for device in data.get('devices', [])]
+            return cls(name, devices)
+        except KeyError as e:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]:
```suggestion
        except (KeyError, TypeError) as e:
```

File: src/pybind/mgr/orchestrator.py [hunk 4]
```diff
@@ -850,6 +856,10 @@ def to_json(self):
                     available=self.available, dev_id=self.dev_id,
                     extended=self.extended)
 
+    @classmethod
+    def from_json(cls, data):
+        return cls(**data)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]: this can raise a TypeError

File: src/pybind/mgr/orchestrator.py (hunk 2 above)
Comment (bk201) [reply src/pybind/mgr/orchestrator.py]: Change to `orchestrator.OrchestratorValidationError`

File: src/pybind/mgr/orchestrator.py (hunk 3 above)
Comment (bk201) [reply src/pybind/mgr/orchestrator.py]: Fixed. Use a decorator to catch type error (https://github.com/ceph/ceph/pull/29595/commits/cee75dc6e27d213a1048f62e0f429a5317d983a4#diff-dfbc76d00f82320b0e6ac732319aa545R665)

Also update unit tests (https://github.com/ceph/ceph/pull/29595/commits/cee75dc6e27d213a1048f62e0f429a5317d983a4#diff-2abf91c18ca9763c937f021625e81599)

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-08-16_15:30:43-rados-wip-kefu-testing-2019-08-16-2103-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-19_10:42:03-rados-wip-kefu-testing-2019-08-16-2103-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-08-20_11:55:18-rados-wip-kefu-testing-2019-08-20-1558-distro-basic-smithi/

**Comment by bk201** [comment]
> * http://pulpito.ceph.com/kchai-2019-08-16_15:30:43-rados-wip-kefu-testing-2019-08-16-2103-distro-basic-mira/
> * http://pulpito.ceph.com/kchai-2019-08-19_10:42:03-rados-wip-kefu-testing-2019-08-16-2103-distro-basic-smithi/
> * http://pulpito.ceph.com/kchai-2019-08-20_11:55:18-rados-wip-kefu-testing-2019-08-20-1558-distro-basic-smithi/

Thanks! Related error log:

```
2019-08-19T11:18:44.054 INFO:teuthology.orchestra.run.smithi033.stderr:Error EINVAL: Traceback (most recent call last):
2019-08-19T11:18:44.054 INFO:teuthology.orchestra.run.smithi033.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 916, in _handle_command
2019-08-19T11:18:44.055 INFO:teuthology.orchestra.run.smithi033.stderr:    return CLICommand.COMMANDS[cmd['prefix']].call(self, cmd, inbuf)
2019-08-19T11:18:44.055 INFO:teuthology.orchestra.run.smithi033.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 368, in call
2019-08-19T11:18:44.055 INFO:teuthology.orchestra.run.smithi033.stderr:    return self.func(mgr, **kwargs)
2019-08-19T11:18:44.055 INFO:teuthology.orchestra.run.smithi033.stderr:  File "/usr/share/ceph/mgr/orchestrator_cli/module.py", line 24, in wrapper
2019-08-19T11:18:44.055 INFO:teuthology.orchestra.run.smithi033.stderr:    return func(*args, **kwargs)
2019-08-19T11:18:44.055 INFO:teuthology.orchestra.run.smithi033.stderr:  File "/usr/share/ceph/mgr/orchestrator_cli/module.py", line 58, in _add_host
2019-08-19T11:18:44.055 INFO:teuthology.orchestra.run.smithi033.stderr:    orchestrator.raise_if_exception(completion)
2019-08-19T11:18:44.056 INFO:teuthology.orchestra.run.smithi033.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 141, in raise_if_exception
2019-08-19T11:18:44.056 INFO:teuthology.orchestra.run.smithi033.stderr:    raise e
2019-08-19T11:18:44.056 INFO:teuthology.orchestra.run.smithi033.stderr:AssertionError
```

Turn out I use the wrong six attribute for string assertion and it causes error in python2 tests.

Originally:

```python
    def add_host(self, host):
        ...
        assert isinstance(host, six.text_type)
```

Change to:
```python
    def add_host(self, host):
        ...
        assert isinstance(host, six.string_types)
```

File: qa/tasks/mgr/test_orchestrator_cli.py [hunk 5]
```diff
@@ -153,3 +159,77 @@ def test_progress(self):
...
+                },
+                {
+                    'nodename': 'host1',
+                    'service_type': 'osd',
+                    'service_instance': '1'
+                }
+            ]
+        }
+
+        self._test_orchestrator_cmd('load_data', '-i', '-', stdin=json.dumps(data))
```
Comment (tchaikov) [inline qa/tasks/mgr/test_orchestrator_cli.py]:
```
2019-08-21T14:40:20.216 INFO:tasks.cephfs_test_runner:======================================================================
2019-08-21T14:40:20.217 INFO:tasks.cephfs_test_runner:ERROR: test_load_data (tasks.mgr.test_orchestrator_cli.TestOrchestratorCli)
2019-08-21T14:40:20.217 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-08-21T14:40:20.217 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-08-21T14:40:20.217 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2019-08-21-1445/qa/tasks/mgr/test_orchestrator_cli.py", line 203, in test_load_data
2019-08-21T14:40:20.217 INFO:tasks.cephfs_test_runner:    self._test_orchestrator_cmd('load_data', '-i', '-', stdin=json.dumps(data))
2019-08-21T14:40:20.217 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2019-08-21-1445/qa/tasks/mgr/test_orchestrator_cli.py", line 30, in _test_orchestrator_cmd
2019-08-21T14:40:20.217 INFO:tasks.cephfs_test_runner:    return self.mgr_cluster.mon_manager.raw_cluster_cmd("test_orchestrator", *args, **kwargs)
2019-08-21T14:40:20.218 INFO:tasks.cephfs_test_runner:TypeError: raw_cluster_cmd() got an unexpected keyword argument 'stdin'
```

see http://pulpito.ceph.com/kchai-2019-08-21_12:58:47-rados-wip-kefu-testing-2019-08-21-1445-distro-basic-mira/4236663

probably you need to update `CephManager.raw_cluster_cmd()` accordingly?

Comment (bk201) [reply qa/tasks/mgr/test_orchestrator_cli.py]: Thanks. Sorry for not spotting this, [raw_cluster_cmd in vstart_runner.py](https://github.com/ceph/ceph/blob/ea216e52f663b04c948f2d43e2736ef2bcb9b04b/qa/tasks/vstart_runner.py#L706) accept kwargs but not in normal [cephfs_test_runner](https://github.com/ceph/ceph/blob/ea216e52f663b04c948f2d43e2736ef2bcb9b04b/qa/tasks/ceph_manager.py#L1229). Update to `raw_cluster_cmd_result`.

**Comment by bk201** [comment]
jenkins retest this please

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-08-23_09:33:33-rados-wip-kefu-testing-2019-08-23-1454-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-25_05:28:04-rados-wip-kefu-testing-2019-08-23-1454-distro-basic-smithi/

**Pull request merged**